File: support/pl/normalizing-lambda/hunting.js
// Redex-hunting

import { parse, show } from './syntax/index.js';

function normalize(term) {
  while (true) {
    const [stepped, madeProgress] = step(term);
    term = stepped;
    if (!madeProgress) {
      return term;
    }
  }
}

// Reduce a single redex if possible. Return an array with 2 elements:
// 1. The (possibly) reduced term.
// 2. `true` if the term was reduced, and `false` otherwise.
function step(term) {
  if (term.type === 'ADDR') {
    return [term, false];
  } else if (term.type === 'FN') {
    const [body, progress] = step(term.body);
    return [{ ...term, body }, progress];
  } else if (term.type === 'APP') {
    if (term.rator.type === 'FN') {
      // Redex.
      // Shift unbound rand addrs _up_ by one in anticipation of being shifted
      // back _down_.
      const arg = shift(term.rand, 1);

      // Substitute arg for the bound addr in the operator's body.
      const afterSubst = subst(term.rator.body, arg);

      // Shift unbound addrs _down_ by 1 to account for the fact that the binder
      // has been removed.
      return [shift(afterSubst, -1), true];
    } else {
      // The operator isn't an abstraction, but it might be reducible to an
      // abstraction (which would reveal a redex on a subsequent step). Try and
      // reduce it.
      const [rator, progress] = step(term.rator);

      if (progress) {
        return [{ ...term, rator }, progress];
      } else {
        // If the operator was already reduced, try and reduce the operand.
        const [rand, progress] = step(term.rand);
        return [{ ...term, rand }, progress];
      }
    }
  }
}

// Shift unbound indexes by `n`.
function shift(term, n, binderCount = 0) {
  if (term.type === 'ADDR') {
    if (term.addr >= binderCount) {
      return {
        ...term,
        addr: term.addr + n,
      };
    } else {
      return term;
    }
  } else if (term.type === 'FN') {
    return {
      ...term,
      body: shift(term.body, n, binderCount + 1),
    };
  } else if (term.type === 'APP') {
    return {
      ...term,
      rator: shift(term.rator, n, binderCount),
      rand: shift(term.rand, n, binderCount),
    };
  }
}

function subst(host, sub, binderCount = 0) {
  if (host.type === 'ADDR') {
    if (host.addr === binderCount) {
      return shift(sub, binderCount);
    } else {
      return host;
    }
  } else if (host.type === 'FN') {
    return {
      ...host,
      body: subst(host.body, sub, binderCount + 1),
    };
  } else if (host.type === 'APP') {
    return {
      ...host,
      rator: subst(host.rator, sub, binderCount),
      rand: subst(host.rand, sub, binderCount),
    };
  }
}

const fact4 = `
(\\self -> \\n -> (n (\\p -> \\t f -> f) (\\t f -> t))
                  (\\s z -> s z)
                  ((\\m n -> m (\\p -> \\s z -> p s (n s z)) (\\s z -> z))
                   n
                   ((self self)
                    ((\\n -> 
                      (n
                       (\\p -> (\\f s -> \\elim -> elim f s)
                               (\\s z -> s ((p \\f s -> f) s z))
                               (p \\f s -> f))
                       ((\\f s -> \\elim -> elim f s)
                        (\\s z -> z)
                        (\\s z -> z)))
                      (\\f s -> s))
                     n))))
(\\self -> \\n -> (n (\\p -> \\t f -> f) (\\t f -> t))
                  (\\s z -> s z)
                  ((\\m n -> m (\\p -> \\s z -> p s (n s z)) (\\s z -> z))
                   n
                   ((self self)
                    ((\\n -> 
                      (n
                       (\\p -> (\\f s -> \\elim -> elim f s)
                               (\\s z -> s ((p \\f s -> f) s z))
                               (p \\f s -> f))
                       ((\\f s -> \\elim -> elim f s)
                        (\\s z -> z)
                        (\\s z -> z)))
                      (\\f s -> s))
                     n))))
(\\s z -> s (s (s (s (s (s z))))))
`;

const normTest = `
(\\x y -> y) ((\\x -> x x) \\x -> x x) (\\x -> x)
`;

const stackTest = `
(\\x -> x x) \\x -> x x
`;

console.log(show(normalize(parse(fact4))));

File: support/pl/normalizing-lambda/Hunting.hs
module Hunting where

data Term = Addr Int
          | Abs String Term
          | App Term Term
  deriving Show

reduce :: Term -> Term
reduce t = case step t of
  (False, t') -> t'
  (True, t')  -> reduce t'

step :: Term -> (Bool, Term)
step t@(Addr _) = (False, t)
step (Abs p b)  = Abs p <$> step b
step (App (Abs _ b) t) = let arg = shift t 1
                             b'  = subst b arg
                          in (True, shift b' (-1))
step (App s t) = (flip App t <$> step s) <||> (App s <$> step t)

shift :: Term -> Int -> Term
shift = loop 0
  where
    loop :: Int -> Term -> Int -> Term
    loop bc t@(Addr a) n
      | a >= bc   = Addr (a + n)
      | otherwise = t
    loop bc (Abs p b) n = Abs p (loop (bc + 1) b n)
    loop bc (App s t) n = App (loop bc s n) (loop bc t n)

subst :: Term -> Term -> Term
subst = loop 0
  where
    loop :: Int -> Term -> Term -> Term
    loop sa t@(Addr a) sub
      | a == sa   = shift sub sa
      | otherwise = t
    loop sa (Abs p b) sub = Abs p (loop (sa + 1) b sub)
    loop sa (App s t) sub = App (loop sa s sub) (loop sa t sub)

(<||>) :: (Bool, a) -> (Bool, a) -> (Bool, a)
t@(True, _) <||> _ = t
_           <||> x = x

infixr 2 <||>

-- (\n -> \s -> \z -> s ((n s) z)) \s -> \z -> s z
one :: Term
one = Abs "s" $ Abs "z" $ App (Addr 1) (Addr 0)

suc :: Term
suc = Abs "n" $ Abs "s" $ Abs "z" $ App (Addr 1) (App (App (Addr 2) (Addr 1)) (Addr 0))

t :: Term
t = App suc one

